File: php/duck-simulator/src/Classes/Duck.php
<?php

namespace App\Classes;

use App\Behavior\FlyBehavior;
use App\Behavior\QuackBehavior;


abstract class Duck {

	/** @var FlyBehavior **/
	private $flyBehavior;
	/** @var QuackBehavior **/	
	private $quackBehavior;

	public function getFlyBehavior(): FlyBehavior {
		return $this->flyBehavior;
	}

	public function setFlyBehavior(FlyBehavior $flyBehavior): void {
		$this->flyBehavior = $flyBehavior;
	}

	public function getQuackBehavior(): QuackBehavior {
		return $this->quackBehavior;
	}

	public function setQuackBehavior(QuackBehavior $quackBehavior): void {
		$this->quackBehavior = $quackBehavior;
	}

	public function fly(): void {
		$this->flyBehavior->fly();
	}

	public function quack(): void {
		$this->quackBehavior->quack();
	}

	abstract public function display(): void;
}

File: python/duck-simulator/src/classes/duck.py
from abc import ABC, abstractmethod

from src.behavior.fly_behavior import FlyBehavior
from src.behavior.quack_behavior import QuackBehavior


class Duck(FlyBehavior, QuackBehavior):
    __metaclass__ = ABC

    __fly_behavior: FlyBehavior
    __quack_behavior: QuackBehavior

    @abstractmethod
    def display(self):
        pass

    def fly(self):
        self.__fly_behavior.fly()

    def quack(self):
        self.__quack_behavior.quack()

    def set_fly_behavior(self, fly_behavior: FlyBehavior):
        self.__fly_behavior = fly_behavior

    def set_quack_behavior(self, quack_behavior: QuackBehavior):
        self.__quack_behavior = quack_behavior
